### PR TITLE
Implement windowing functions without using grammar rules

### PR DESCRIPTION
**Pull request by jimidle**
Windowing functions were all over specified and can actually be built from the visitExprOver, which this PR achieves.

Note that the existing ir for this seems somewhat under specified in that it does not use Option and does not define placeholders for arguments that will be irrelevant. 

progresses: #357 
progresses: #275

### CODE REVIEW

**Comment by vil1** [comment]
> Windowing functions were all over specified and can actually be built from the visitExprOver, which this PR achieves.

Windowing functions being `OVER` specified is no surprise 😆

File: core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlExpressionBuilder.scala [hunk 1]
```diff
@@ -3,9 +3,10 @@ package com.databricks.labs.remorph.parsers.tsql
 import com.databricks.labs.remorph.parsers.tsql.TSqlParser._
 import com.databricks.labs.remorph.parsers.{FunctionBuilder, ParserCommon, intermediate => ir}
 import org.antlr.v4.runtime.Token
-import org.antlr.v4.runtime.tree.TerminalNode
+import org.antlr.v4.runtime.tree.{TerminalNode, Trees}
 
 import scala.collection.JavaConverters.asScalaBufferConverter
+import scala.collection.convert.ImplicitConversions.`collection AsScalaIterable`
```
Comment (vil1) [inline core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlExpressionBuilder.scala:9]:
```suggestion
import scala.collection.JavaConverters._
```

File: core/src/main/scala/com/databricks/labs/remorph/parsers/intermediate/extensions.scala [hunk 2]
```diff
@@ -150,3 +150,5 @@ case class ArrayAccess(array: Expression, index: Expression) extends Expression
 case class JsonAccess(json: Expression, path: Seq[String]) extends Expression {}
 case class Collate(string: Expression, specification: String) extends Expression {}
 case class Iff(condition: Expression, thenBranch: Expression, elseBranch: Expression) extends Expression {}
+
+case class Over(input: Expression, partitionBy: Seq[Expression], orderBy: Seq[SortOrder]) extends Expression {}
```
Comment (vil1) [inline core/src/main/scala/com/databricks/labs/remorph/parsers/intermediate/extensions.scala:154]: Seems unused 

```suggestion
```

File: core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlExpressionBuilder.scala [hunk 3]
```diff
@@ -190,4 +191,73 @@ class TSqlExpressionBuilder extends TSqlParserBaseVisitor[ir.Expression] with Pa
...
+      case Some(frameStartNode: WindowFrameBoundContext) =>
+        val frameStart = buildFrame(frameStartNode)
+        val frameEnd = fs.tail.headOption match {
+          case Some(frameEndNode: WindowFrameBoundContext) => buildFrame(frameEndNode)
+          case _ => ir.FrameBoundary(current_row = false, unbounded = false, ir.Noop)
+        }
+        ir.WindowFrame(frameType, frameStart, frameEnd)
+      case _ => // Cannot reach
+        noWindowFrame
+    }
```
Comment (vil1) [inline core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlExpressionBuilder.scala:240]: This part could be extracted as a separate`private [tsql]` function taking a `FrameType` and a `Collection[ParseTree]` (or some scala counterpart of `Collection[ParseTree]`), making it amenable to unit test covering the unreachable state (using mockito).

```suggestion
    fs.headOption match {
      case Some(frameStartNode: WindowFrameBoundContext) =>
        val frameStart = buildFrame(frameStartNode)
        val frameEnd = fs.tail.headOption match {
          case Some(frameEndNode: WindowFrameBoundContext) => buildFrame(frameEndNode)
          case _ => ir.FrameBoundary(current_row = false, unbounded = false, ir.Noop)
        }
        ir.WindowFrame(frameType, frameStart, frameEnd)
      case _ => // Cannot reach
        noWindowFrame
    }
```

**Review by vil1 — CHANGES_REQUESTED** [review]
Needs some slight adjustments

File: core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlExpressionBuilder.scala (hunk 1 above)
Comment (jimidle) [reply core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlExpressionBuilder.scala]: headOption cannot be used with that change. It seems the equivalent in JavaConverters is deprecated according to IntelliJ Ultimate at least.

File: core/src/main/scala/com/databricks/labs/remorph/parsers/intermediate/extensions.scala (hunk 2 above)
Comment (jimidle) [reply core/src/main/scala/com/databricks/labs/remorph/parsers/intermediate/extensions.scala:154]: Ah yes. It was how I was going to do it until I saw that there was existing Ir.

File: core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlExpressionBuilder.scala (hunk 3 above)
Comment (jimidle) [reply core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlExpressionBuilder.scala]: Reworked as new function that avoids redundant case altogether

File: core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlExpressionBuilder.scala [hunk 4]
```diff
@@ -224,6 +223,73 @@ class TSqlExpressionBuilder extends TSqlParserBaseVisitor[ir.Expression] with Pa
...
+    val frameType = buildFrameType(ctx)
+    val bounds = Trees
+      .findAllRuleNodes(ctx, TSqlParser.RULE_windowFrameBound)
+      .asScala
+      .collect { case wfb: WindowFrameBoundContext => wfb }
+      .map(buildFrame)
+
+    val frameStart = bounds.head // Safe due to the nature of window frames always having at least a start bound
+    val frameEnd =
+      bounds.drop(1).headOption.getOrElse(ir.FrameBoundary(current_row = false, unbounded = false, ir.Noop))
```
Comment (vil1) [inline core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlExpressionBuilder.scala:267]: nitpicking: `bounds.tail...` would achieve the same with fewer keystrokes.

File: core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlExpressionBuilder.scala [hunk 5]
```diff
@@ -224,6 +223,73 @@ class TSqlExpressionBuilder extends TSqlParserBaseVisitor[ir.Expression] with Pa
...
+      ir.FrameBoundary(current_row = true, unbounded = false, ir.Noop)
+    } else {
+      Option(ctx.INT())
+        .map(intNode =>
+          ir.FrameBoundary(
+            current_row = false,
+            unbounded = false,
+            value = ir.Literal(integer = Some(intNode.getText.toInt))))
+        .getOrElse(ir.FrameBoundary(current_row = false, unbounded = false, ir.Noop))
+    }
```
Comment (vil1) [inline core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlExpressionBuilder.scala:290]: For consistency with the rest of the code:

```suggestion
    ctx match {
      case c if c.UNBOUNDED() != null => r.FrameBoundary(current_row = false, unbounded = true, value = ir.Noop)
      case c if c.CURRENT() != null => ir.FrameBoundary(current_row = true, unbounded = false, ir.Noop)
      case c if c.INT() != null =>
         ir.FrameBoundary(
            current_row = false,
            unbounded = false,
            value = ir.Literal(integer = Some(c.INT().getText.toInt)))
      case _ => ir.FrameBoundary(current_row = false, unbounded = false, ir.Noop)
    }
```

**Review by vil1 — CHANGES_REQUESTED** [review]
One last change request

File: core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlExpressionBuilder.scala (hunk 4 above)
Comment (jimidle) [reply core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlExpressionBuilder.scala:267]: Cool.

File: core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlExpressionBuilder.scala (hunk 5 above)
Comment (jimidle) [reply core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlExpressionBuilder.scala:290]: OK - no problem, (well, without the typo)

**Comment by jimidle** [comment]
@sundarshankar89 - Can we add this to the merge queue now please?

**Review by vil1 — APPROVED** [review]
LGTM

**Comment by sundarshankar89** [comment]
@jimidle you need fix the codecoverage, I can add it but it will only be merged once you fix the code coverage

**Comment by jimidle** [comment]
> @jimidle you need fix the codecoverage, I can add it but it will only be merged once you fix the code coverage

OK - this is where code coverage is a little lame.

**Pull request merged**